## 데이터셋 준비
- The Complete Titanic Dataset : https://www.kaggle.com/datasets/vinicius150987/titanic3
- 다양한 타이타닉 Dataset : https://www.kaggle.com/datasets?search=titanic

In [1]:
%%html
<style> .rendered_html table{margin-left:5px} </style>

In [2]:
import pandas as pd

In [4]:
ttd=pd.read_excel('titanic3.xls')
ttd

pclass  survived                                             name  \
0          1         1                    Allen, Miss. Elisabeth Walton   
1          1         1                   Allison, Master. Hudson Trevor   
2          1         0                     Allison, Miss. Helen Loraine   
3          1         0             Allison, Mr. Hudson Joshua Creighton   
4          1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...      ...       ...                                              ...   
1304       3         0                             Zabour, Miss. Hileni   
1305       3         0                            Zabour, Miss. Thamine   
1306       3         0                        Zakarian, Mr. Mapriededer   
1307       3         0                              Zakarian, Mr. Ortin   
1308       3         0                               Zimmerman, Mr. Leo   

         sex      age  sibsp  parch  ticket      fare    cabin embarked boat  \
0     female  29.0000      0      0   24160  211.3375       B5        S    2   
1       male   0.9167      1      2  113781  151.5500  C22 C26        S   11   
2     female   2.0000      1      2  113781  151.5500  C22 C26        S  NaN   
3       male  30.0000      1      2  113781  151.5500  C22 C26        S  NaN   
4     female  25.0000      1      2  113781  151.5500  C22 C26        S  NaN   
...      ...      ...    ...    ...     ...       ...      ...      ...  ...   
1304  female  14.5000      1      0    2665   14.4542      NaN        C  NaN   
1305  female      NaN      1      0    2665   14.4542      NaN        C  NaN   
1306    male  26.5000      0      0    2656    7.2250      NaN        C  NaN   
1307    male  27.0000      0      0    2670    7.2250      NaN        C  NaN   
1308    male  29.0000      0      0  315082    7.8750      NaN        S  NaN   

       body                        home.dest  
0       NaN                     St Louis, MO  
1       NaN  Montreal, PQ / Chesterville, ON  
2       NaN  Montreal, PQ / Chesterville, ON  
3     135.0  Montreal, PQ / Chesterville, ON  
4       NaN  Montreal, PQ / Chesterville, ON  
...     ...                              ...  
1304  328.0                              NaN  
1305    NaN                              NaN  
1306  304.0                              NaN  
1307    NaN                              NaN  
1308    NaN                              NaN  

[1309 rows x 14 columns]

만일 다음과 같은 에러가 발생한다면 xlrd설치가 필요하다.  
`Missing optional dependency 'xlrd'. Install xlrd >= 1.0.0 for Excel support Use pip or conda to install xlrd.`

### 1. About Dataset
타이타닉 데이터셋은 타이타닉호의 침몰 사건과 관련된 정보를 담고 있습니다.   
타이타닉호는 역사상 가장 유명한 난파선 사건 중 하나입니다.

1912년 4월 15일, 타이타닉호는 첫 항해 중에 있던 도중에 빙산과 충돌하여 침몰했습니다.  
이 배는 "침몰하지 않을 것"이라고 여겨져 왔지만,  
탑승자들을 위한 구명보트가 충분하지 않아 2224명의 승객과 선원 중 1502명이 목숨을 잃었습니다.

생존에는 일부 운이 개입되었지만,  
특정 그룹의 사람들은 다른 사람들보다 생존할 가능성이 더 높은 것으로 보입니다.

이 과제에서는 **승객들의 데이터(예: 이름, 나이, 성별, 사회경제적 지위 등)** 를 사용하여  
**"어떤 종류의 사람들이 생존할 가능성이 더 높았는가?"** 라는 예측 모델을 구축하도록 요청합니다.

### 2. 예상해볼 수 있는 분석
**생존율 분석**  
승객들의 **특성(성별, 나이, 객실 등급 등)에 따른 생존율을 분석**할 수 있습니다.  
어떤 그룹의 사람들이 생존할 가능성이 높았는지, **그들의 특징은 무엇인지** 알 수 있습니다.

**특성 엔지니어링**  
데이터셋에는 승객의 이름, 성별, 나이, 객실 등급, 승선 항구 등 다양한 특성이 포함되어 있습니다.  
이러한 특성들을 활용하여 **새로운 특성을 만들어내거나 범주형 특성을 수치형으로 변환**할 수 있습니다.

**데이터 시각화**  
데이터셋을 시각화하여 **생존 여부와 관련된 패턴이나 상관관계를 탐색**할 수 있습니다.  
막대 그래프, 히스토그램, 상자 그림, 히트맵 등 다양한 시각화 기법을 사용하여 데이터를 탐색하고 이해할 수 있습니다.

**예측 모델링**  
타이타닉 데이터셋을 사용하여 생존 여부를 예측하는 머신 러닝 모델을 개발할 수 있습니다.  
다양한 알고리즘을 사용하여 모델을 학습하고 테스트하여 최적의 모델을 찾을 수 있습니다.

**특성 중요도 분석**  
학습된 모델을 통해 **각 특성의 중요도**를 분석할 수 있습니다.  
어떤 특성이 생존 예측에 가장 중요한 역할을 하는지 파악할 수 있습니다.

### 3. 항목 설명
항목 | 설명
-|:-
pclass	| Passenger Class (1 = 1st; 2 = 2nd; 3 = 3rd), 승객의 등급 (1 = 1등급, 2 = 2등급, 3 = 3등급)
survived | Survival (0 = No; 1 = Yes), 생존 여부 (0 = 생존하지 않음, 1 = 생존)
name | Name, 이름
sex	| Sex, 성별
age | Age, 나이
sibsp | Number of Siblings/Spouses Aboard, 함께 탑승한 형제자매/배우자의 수
parch | Number of Parents/Children Aboard, 함께 탑승한 부모/자녀의 수
ticket | Ticket Number, 티켓 번호
fare | Passenger Fare (British pound), 승객 요금 (영국 파운드)
cabin | Cabin, 객실 번호
embarked | Port of Embarkation (C = Cherbourg; Q = Queenstown; S = Southampton), 탑승 항구 (C = 쉐르부르, Q = 퀸스타운, S = 사우샘프턴)
boat | Lifeboat, 구명보트 번호
body | Body Identification Number, 시신 식별 번호
home.dest | Home/Destination, 거주지/목적지

### 4. 특이사항
**Pclass는 사회경제적 지위(SES)를 나타내는 대리 변수입니다.**   
1등급은 상류층, 2등급은 중산층, 3등급은 하류층을 대표합니다.  
  
**Age는 연령으로 표시되며, 만약 1보다 적은 경우 소수점으로 표현됩니다.**    
만약 연령이 추정된 경우 xx.5와 같은 형식으로 표시됩니다.
  
**Fare는 1970년 이전의 영국 파운드로 표시됩니다.**  
-환산 계수: 1 파운드 = 12 실링 = 240 페니, 1 실링 = 20 페니
  
**가족 관계 변수인 sibsp와 parch에 대해서는 일부 관계가 무시되었습니다.**   
-다음은 sibsp와 parch에 사용된 정의입니다.  
-Sibling: 타이타닉호에 함께 탑승한 형제, 자매, 이복형제, 이복자매  
-Spouse: 타이타닉호에 함께 탑승한 남편 또는 아내 (연인과 약혼자는 제외)  
-Parent: 타이타닉호에 함께 탑승한 어머니 또는 아버지  
-Child: 타이타닉호에 함께 탑승한 아들, 딸, 이복아들, 이복딸  
  
**이 연구에서 제외된 다른 가족 관계에는 사촌, 조카, 이모, 고모, 시부모 등이 포함됩니다.**   
-일부 아이들은 유모와 함께만 여행했으므로, 그들에 대해서는 parch=0으로 표시되었습니다.   
-또한, 일부 사람들은 아주 가까운 친구나 이웃과 함께 여행했지만, 이러한 관계는 정의에서 지원되지 않습니다.

## 5. dtype 등 info 파악

In [7]:
# 타이타닉 데이터셋 info
ttd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   pclass     1309 non-null   int64  
 1   survived   1309 non-null   int64  
 2   name       1309 non-null   object 
 3   sex        1309 non-null   object 
 4   age        1046 non-null   float64
 5   sibsp      1309 non-null   int64  
 6   parch      1309 non-null   int64  
 7   ticket     1309 non-null   object 
 8   fare       1308 non-null   float64
 9   cabin      295 non-null    object 
 10  embarked   1307 non-null   object 
 11  boat       486 non-null    object 
 12  body       121 non-null    float64
 13  home.dest  745 non-null    object 
dtypes: float64(3), int64(4), object(7)
memory usage: 143.3+ KB


1. 데이터 행은 총 1309개
2. pclass, survived, name, sex, sibsp, parch, ticket 은 결측치 없음
3. age, fare, cabin, embarked, boat, body, home.dest는 결측치 있음.
4. int64 : pclass, survived, sibsp, parch
5. float64 : age, fare, body
6. object : name, sex, ticket, cabin, embarked, boat, home.dest

## 데이터프레임 준비
- 실습을 위해 name을 인덱스로 사용할 예정

In [11]:
# index_col = 'name'으로 지정
ttd=pd.read_excel('titanic3.xls', index_col='name')
ttd.head()

pclass  survived     sex  \
name                                                                        
Allen, Miss. Elisabeth Walton                         1         1  female   
Allison, Master. Hudson Trevor                        1         1    male   
Allison, Miss. Helen Loraine                          1         0  female   
Allison, Mr. Hudson Joshua Creighton                  1         0    male   
Allison, Mrs. Hudson J C (Bessie Waldo Daniels)       1         0  female   

                                                     age  sibsp  parch  \
name                                                                     
Allen, Miss. Elisabeth Walton                    29.0000      0      0   
Allison, Master. Hudson Trevor                    0.9167      1      2   
Allison, Miss. Helen Loraine                      2.0000      1      2   
Allison, Mr. Hudson Joshua Creighton             30.0000      1      2   
Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  25.0000      1      2   

                                                 ticket      fare    cabin  \
name                                                                         
Allen, Miss. Elisabeth Walton                     24160  211.3375       B5   
Allison, Master. Hudson Trevor                   113781  151.5500  C22 C26   
Allison, Miss. Helen Loraine                     113781  151.5500  C22 C26   
Allison, Mr. Hudson Joshua Creighton             113781  151.5500  C22 C26   
Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  113781  151.5500  C22 C26   

                                                embarked boat   body  \
name                                                                   
Allen, Miss. Elisabeth Walton                          S    2    NaN   
Allison, Master. Hudson Trevor                         S   11    NaN   
Allison, Miss. Helen Loraine                           S  NaN    NaN   
Allison, Mr. Hudson Joshua Creighton                   S  NaN  135.0   
Allison, Mrs. Hudson J C (Bessie Waldo Daniels)        S  NaN    NaN   

                                                                       home.dest  
name                                                                              
Allen, Miss. Elisabeth Walton                                       St Louis, MO  
Allison, Master. Hudson Trevor                   Montreal, PQ / Chesterville, ON  
Allison, Miss. Helen Loraine                     Montreal, PQ / Chesterville, ON  
Allison, Mr. Hudson Joshua Creighton             Montreal, PQ / Chesterville, ON  
Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  Montreal, PQ / Chesterville, ON

In [19]:
# 정렬 : 이름 순
# 문자부호표 원리로 소문자가 뒤로 배치 됨
ttd=ttd.sort_index()
ttd

pclass  survived     sex   age  \
name                                                                            
Abbing, Mr. Anthony                                 3         0    male  42.0   
Abbott, Master. Eugene Joseph                       3         0    male  13.0   
Abbott, Mr. Rossmore Edward                         3         0    male  16.0   
Abbott, Mrs. Stanton (Rosa Hunt)                    3         1  female  35.0   
Abelseth, Miss. Karen Marie                         3         1  female  16.0   
...                                               ...       ...     ...   ...   
del Carlo, Mrs. Sebastiano (Argenia Genovesi)       2         1  female  24.0   
van Billiard, Master. James William                 3         0    male   NaN   
van Billiard, Master. Walter John                   3         0    male  11.5   
van Billiard, Mr. Austin Blyler                     3         0    male  40.5   
van Melkebeke, Mr. Philemon                         3         0    male   NaN   

                                               sibsp  parch         ticket  \
name                                                                         
Abbing, Mr. Anthony                                0      0      C.A. 5547   
Abbott, Master. Eugene Joseph                      0      2      C.A. 2673   
Abbott, Mr. Rossmore Edward                        1      1      C.A. 2673   
Abbott, Mrs. Stanton (Rosa Hunt)                   1      1      C.A. 2673   
Abelseth, Miss. Karen Marie                        0      0         348125   
...                                              ...    ...            ...   
del Carlo, Mrs. Sebastiano (Argenia Genovesi)      1      0  SC/PARIS 2167   
van Billiard, Master. James William                1      1       A/5. 851   
van Billiard, Master. Walter John                  1      1       A/5. 851   
van Billiard, Mr. Austin Blyler                    0      2       A/5. 851   
van Melkebeke, Mr. Philemon                        0      0         345777   

                                                  fare cabin embarked boat  \
name                                                                         
Abbing, Mr. Anthony                             7.5500   NaN        S  NaN   
Abbott, Master. Eugene Joseph                  20.2500   NaN        S  NaN   
Abbott, Mr. Rossmore Edward                    20.2500   NaN        S  NaN   
Abbott, Mrs. Stanton (Rosa Hunt)               20.2500   NaN        S    A   
Abelseth, Miss. Karen Marie                     7.6500   NaN        S   16   
...                                                ...   ...      ...  ...   
del Carlo, Mrs. Sebastiano (Argenia Genovesi)  27.7208   NaN        C   12   
van Billiard, Master. James William            14.5000   NaN        S  NaN   
van Billiard, Master. Walter John              14.5000   NaN        S  NaN   
van Billiard, Mr. Austin Blyler                14.5000   NaN        S  NaN   
van Melkebeke, Mr. Philemon                     9.5000   NaN        S  NaN   

                                                body  \
name                                                   
Abbing, Mr. Anthony                              NaN   
Abbott, Master. Eugene Joseph                    NaN   
Abbott, Mr. Rossmore Edward                    190.0   
Abbott, Mrs. Stanton (Rosa Hunt)                 NaN   
Abelseth, Miss. Karen Marie                      NaN   
...                                              ...   
del Carlo, Mrs. Sebastiano (Argenia Genovesi)    NaN   
van Billiard, Master. James William              NaN   
van Billiard, Master. Walter John                1.0   
van Billiard, Mr. Austin Blyler                255.0   
van Melkebeke, Mr. Philemon                      NaN   

                                                               home.dest  
name                                                                      
Abbing, Mr. Anthony                                                  NaN  
Abbott, M

In [23]:
# 정렬 : 이름 순 - 대소문자 상관없이 정렬
# key : 크고 작음을 처리하는 방식을 정하는 함수 지정
ttd=ttd.sort_index(key= lambda x: x.str.lower())
ttd

pclass  survived     sex   age  sibsp  \
name                                                                      
Abbing, Mr. Anthony                    3         0    male  42.0      0   
Abbott, Master. Eugene Joseph          3         0    male  13.0      0   
Abbott, Mr. Rossmore Edward            3         0    male  16.0      1   
Abbott, Mrs. Stanton (Rosa Hunt)       3         1  female  35.0      1   
Abelseth, Miss. Karen Marie            3         1  female  16.0      0   
...                                  ...       ...     ...   ...    ...   
Zabour, Miss. Hileni                   3         0  female  14.5      1   
Zabour, Miss. Thamine                  3         0  female   NaN      1   
Zakarian, Mr. Mapriededer              3         0    male  26.5      0   
Zakarian, Mr. Ortin                    3         0    male  27.0      0   
Zimmerman, Mr. Leo                     3         0    male  29.0      0   

                                  parch     ticket     fare cabin embarked  \
name                                                                         
Abbing, Mr. Anthony                   0  C.A. 5547   7.5500   NaN        S   
Abbott, Master. Eugene Joseph         2  C.A. 2673  20.2500   NaN        S   
Abbott, Mr. Rossmore Edward           1  C.A. 2673  20.2500   NaN        S   
Abbott, Mrs. Stanton (Rosa Hunt)      1  C.A. 2673  20.2500   NaN        S   
Abelseth, Miss. Karen Marie           0     348125   7.6500   NaN        S   
...                                 ...        ...      ...   ...      ...   
Zabour, Miss. Hileni                  0       2665  14.4542   NaN        C   
Zabour, Miss. Thamine                 0       2665  14.4542   NaN        C   
Zakarian, Mr. Mapriededer             0       2656   7.2250   NaN        C   
Zakarian, Mr. Ortin                   0       2670   7.2250   NaN        C   
Zimmerman, Mr. Leo                    0     315082   7.8750   NaN        S   

                                 boat   body               home.dest  
name                                                                  
Abbing, Mr. Anthony               NaN    NaN                     NaN  
Abbott, Master. Eugene Joseph     NaN    NaN     East Providence, RI  
Abbott, Mr. Rossmore Edward       NaN  190.0     East Providence, RI  
Abbott, Mrs. Stanton (Rosa Hunt)    A    NaN     East Providence, RI  
Abelseth, Miss. Karen Marie        16    NaN  Norway Los Angeles, CA  
...                               ...    ...                     ...  
Zabour, Miss. Hileni              NaN  328.0                     NaN  
Zabour, Miss. Thamine             NaN    NaN                     NaN  
Zakarian, Mr. Mapriededer         NaN  304.0                     NaN  
Zakarian, Mr. Ortin               NaN    NaN                     NaN  
Zimmerman, Mr. Leo                NaN    NaN                     NaN  

[1309 rows x 13 columns]

## .loc
- https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.loc.html
- 인덱스 레이블로 행 조회
- 메소드(함수)가 아닌 속성(property)
- `.loc[]`은 주로 레이블(label)을 기반으로 행과 열에 접근하는 데 사용되지만, boolean 배열과 함께 사용할 수도 있습니다.

**다음과 같은 입력이 가능합니다**
1. 하나의 레이블
    - `5` 또는 `'a'` (5는 인덱스의 레이블로 해석되며 정수 위치로 해석되지는 않음).
2. 레이블의 리스트 또는 배열
    - `['a', 'b', 'c']`
3. 레이블 범위로 된 슬라이스 객체
    - `'a':'f'`
4. 축과 길이가 같은 boolean 배열
    - `[True, False, True]`

In [25]:
ttd.loc

In [27]:
# 1. Abbing, Mr. Anthony 를 조회
# 5 또는 'a'
ttd.loc['Abbing, Mr. Anthony']

pclass               3
survived             0
sex               male
age               42.0
sibsp                0
parch                0
ticket       C.A. 5547
fare              7.55
cabin              NaN
embarked             S
boat               NaN
body               NaN
home.dest          NaN
Name: Abbing, Mr. Anthony, dtype: object

In [29]:
# 1-1 시리즈 말고 데이터프레임으로 받고 싶으면
ttd.loc[['Abbing, Mr. Anthony']]

pclass  survived   sex   age  sibsp  parch     ticket  \
name                                                                         
Abbing, Mr. Anthony       3         0  male  42.0      0      0  C.A. 5547   

                     fare cabin embarked boat  body home.dest  
name                                                           
Abbing, Mr. Anthony  7.55   NaN        S  NaN   NaN       NaN

In [31]:
# 2. Abbing, Mr. Anthony와 Zimmerman, Mr. Leo 를 조회
# ['a', 'b', 'c']
ttd.loc[['Abbing, Mr. Anthony','Zimmerman, Mr. Leo']]

pclass  survived   sex   age  sibsp  parch     ticket  \
name                                                                         
Abbing, Mr. Anthony       3         0  male  42.0      0      0  C.A. 5547   
Zimmerman, Mr. Leo        3         0  male  29.0      0      0     315082   

                      fare cabin embarked boat  body home.dest  
name                                                            
Abbing, Mr. Anthony  7.550   NaN        S  NaN   NaN       NaN  
Zimmerman, Mr. Leo   7.875   NaN        S  NaN   NaN       NaN

In [33]:
# 3. Abbott, Master. Eugene Joseph 에서 Abelseth, Miss. Karen Marie까지 조회
# 'a':'f'
ttd.loc['Abbott, Master. Eugene Joseph':'Abelseth, Miss. Karen Marie']

pclass  survived     sex   age  sibsp  \
name                                                                      
Abbott, Master. Eugene Joseph          3         0    male  13.0      0   
Abbott, Mr. Rossmore Edward            3         0    male  16.0      1   
Abbott, Mrs. Stanton (Rosa Hunt)       3         1  female  35.0      1   
Abelseth, Miss. Karen Marie            3         1  female  16.0      0   

                                  parch     ticket   fare cabin embarked boat  \
name                                                                            
Abbott, Master. Eugene Joseph         2  C.A. 2673  20.25   NaN        S  NaN   
Abbott, Mr. Rossmore Edward           1  C.A. 2673  20.25   NaN        S  NaN   
Abbott, Mrs. Stanton (Rosa Hunt)      1  C.A. 2673  20.25   NaN        S    A   
Abelseth, Miss. Karen Marie           0     348125   7.65   NaN        S   16   

                                   body               home.dest  
name                                                             
Abbott, Master. Eugene Joseph       NaN     East Providence, RI  
Abbott, Mr. Rossmore Edward       190.0     East Providence, RI  
Abbott, Mrs. Stanton (Rosa Hunt)    NaN     East Providence, RI  
Abelseth, Miss. Karen Marie         NaN  Norway Los Angeles, CA

In [37]:
ttd['Abbott, Master. Eugene Joseph':'Abelseth, Miss. Karen Marie']

pclass  survived     sex   age  sibsp  \
name                                                                      
Abbott, Master. Eugene Joseph          3         0    male  13.0      0   
Abbott, Mr. Rossmore Edward            3         0    male  16.0      1   
Abbott, Mrs. Stanton (Rosa Hunt)       3         1  female  35.0      1   
Abelseth, Miss. Karen Marie            3         1  female  16.0      0   

                                  parch     ticket   fare cabin embarked boat  \
name                                                                            
Abbott, Master. Eugene Joseph         2  C.A. 2673  20.25   NaN        S  NaN   
Abbott, Mr. Rossmore Edward           1  C.A. 2673  20.25   NaN        S  NaN   
Abbott, Mrs. Stanton (Rosa Hunt)      1  C.A. 2673  20.25   NaN        S    A   
Abelseth, Miss. Karen Marie           0     348125   7.65   NaN        S   16   

                                   body               home.dest  
name                                                             
Abbott, Master. Eugene Joseph       NaN     East Providence, RI  
Abbott, Mr. Rossmore Edward       190.0     East Providence, RI  
Abbott, Mrs. Stanton (Rosa Hunt)    NaN     East Providence, RI  
Abelseth, Miss. Karen Marie         NaN  Norway Los Angeles, CA

In [39]:
# 4. 없는 이름 조회
ttd.loc['s']

KeyError: 's'

### 없는 경우에 대처 방법 - if..else..

In [ ]:
# if문으로 제어
if 'hee' in ttd.index:
    result=tt.loc['hee']
else:
    result = None # 또는 다른 조치

### 없는 경우에 대처 방법 - try..except..

In [ ]:
# try문으로 제어
try:
    result=tt.loc['hee']
except:
    result=None 

## .duplicated() - 동명이인 발생하는 경우
- 정리한 스프레드시트 : https://docs.google.com/spreadsheets/d/1p8SgoOqLPGZbkkNBfWDyxdjlM1sBoOLo/edit?usp=sharing&ouid=104295259370579267197&rtpof=true&sd=true
- 동명 이인
    - 'Connolly, Miss. Kate'
    - 'Kelly, Mr. James'

- DataFrame.index : https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.index.html
- DataFrame.duplicated : https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.duplicated.html

In [41]:
# 'Connolly, Miss. Kate' 조회
ttd.loc['Connolly, Miss. Kate']

pclass  survived     sex   age  sibsp  parch  ticket  \
name                                                                         
Connolly, Miss. Kate       3         1  female  22.0      0      0  370373   
Connolly, Miss. Kate       3         0  female  30.0      0      0  330972   

                        fare cabin embarked boat  body home.dest  
name                                                              
Connolly, Miss. Kate  7.7500   NaN        Q   13   NaN   Ireland  
Connolly, Miss. Kate  7.6292   NaN        Q  NaN   NaN   Ireland

In [43]:
# 'Kelly, Mr. James'
ttd.loc['Kelly, Mr. James']

pclass  survived   sex   age  sibsp  parch  ticket    fare  \
name                                                                           
Kelly, Mr. James       3         0  male  34.5      0      0  330911  7.8292   
Kelly, Mr. James       3         0  male  44.0      0      0  363592  8.0500   

                 cabin embarked boat  body home.dest  
name                                                  
Kelly, Mr. James   NaN        Q  NaN  70.0       NaN  
Kelly, Mr. James   NaN        S  NaN   NaN       NaN

In [45]:
# 1. 데이터프레임.인덱스.중복확인()
ttd.index.duplicated()

array([False, False, False, ..., False, False, False])

In [49]:
# 2. 1에 조건에 맞는 행 출력
ttd.loc[ttd.index.duplicated()]

pclass  survived     sex   age  sibsp  parch  ticket  \
name                                                                         
Connolly, Miss. Kate       3         0  female  30.0      0      0  330972   
Kelly, Mr. James           3         0    male  44.0      0      0  363592   

                        fare cabin embarked boat  body home.dest  
name                                                              
Connolly, Miss. Kate  7.6292   NaN        Q  NaN   NaN   Ireland  
Kelly, Mr. James      8.0500   NaN        S  NaN   NaN       NaN

총 4명이 나와야 하는데, 하나의 데이터만 나오고 있다.

In [51]:
# 여기서부터 이미 2만 나오고 있던 상황
ttd.index.duplicated().sum()

2

## keep=

**keep : 중복된 값 중 어떤 값을 유지할지를 지정**
1. first
    - 첫 번째 등장하는 값을 제외하고 중복된 값을 True로 표시
    - 중복된 값 중 첫 번째 값은 유지되고, 이후 등장하는 중복된 값들은 True로 표시.
2. last
    - 마지막 등장하는 값을 제외하고 중복된 값을 True로 표시
    - 중복된 값 중 마지막 값은 유지되고, 이전에 등장하는 중복된 값들은 True로 표시.
3. False 
    - 모든 중복된 값을 True로 표시합니다
    - 중복된 값들 모두를 유지하지 않고 True로 표시.

In [53]:
# 3. 2에서 중복값을 포함하도록 한다.
ttd.loc[ttd.index.duplicated(keep=False)]

pclass  survived     sex   age  sibsp  parch  ticket  \
name                                                                         
Connolly, Miss. Kate       3         1  female  22.0      0      0  370373   
Connolly, Miss. Kate       3         0  female  30.0      0      0  330972   
Kelly, Mr. James           3         0    male  34.5      0      0  330911   
Kelly, Mr. James           3         0    male  44.0      0      0  363592   

                        fare cabin embarked boat  body home.dest  
name                                                              
Connolly, Miss. Kate  7.7500   NaN        Q   13   NaN   Ireland  
Connolly, Miss. Kate  7.6292   NaN        Q  NaN   NaN   Ireland  
Kelly, Mr. James      7.8292   NaN        Q  NaN  70.0       NaN  
Kelly, Mr. James      8.0500   NaN        S  NaN   NaN       NaN